Sources/SwiftSyntaxMacrosTestSupport/Assertions.swift

```diff
@@ -254,13 +254,17 @@ func assertDiagnostic(
 ///   - macros: The macros that should be expanded, provided as a dictionary
 ///     mapping macro names (e.g., `"stringify"`) to implementation types
 ///     (e.g., `StringifyMacro.self`).
+///   - conformanceMap: The protocols that macro should conform to, provided as
+///     a dictionary mapping macro names (e.g., `"AddCodable"`) to protocols names
+///     (e.g., `"Decodable"`, `"Encodable"` etc.).
 ///   - testModuleName: The name of the test module to use.
 ///   - testFileName: The name of the test file name to use.
 public func assertMacroExpansion(
   _ originalSource: String,
   expandedSource expectedExpandedSource: String,
   diagnostics: [DiagnosticSpec] = [],
   macros: [String: Macro.Type],
+  conformsTo conformanceMap: [String: [TypeSyntax]] = [:],
```

Comment:
I’m not a fan of adding another parameter to `assertMacroExpansion`. What I would prefer is that the `macros` parameter becomes `[String: MacroSpec]` with 

```swift
struct MacroSpec {
  let type: MacroType
  let conformances: [TypeSyntax]
}
```

The benefit of that is that we could expand `MacroSpec` in the future without the need to have more parameters to `assertMacroExpansion`.

We still need to keep the current function that takes `macros: [String: Macro.Type]` around for API-compatibility.

Reply:
The concept of a `[String: MacroSpec]` dictionary is neat, and I think we can build on that to create an even more robust and clear model.

How about structuring `MacroSpec` to encapsulate parameters specific to each macro's role? We could use an enum with associated values for this purpose, which would provide us with both flexibility and clarity:

```swift
enum MacroRoleContext {
  case expression
  case declaration
  case accessor
  case memberAttribute
  case member
  case peer
  case conformance
  case codeItem
  case `extension`(conformances: [TypeSyntax])
}

struct MacroSpec {
  let type: MacroType
  let macroContext: MacroRoleContext?
}
```

With this design, `MacroSpec` could serve as a clear and intuitive representation of our macro configurations. It inherently guides the developer to provide only the relevant context for each macro role, which enhances code readability and makes future expansions or modifications to `MacroSpec` more manageable.

Reply:
This does sound very compelling but every change to add new members to one of the macro roles would be API-breaking. For example, recently the `member` macro role gained `conformances` as well (https://github.com/apple/swift-syntax/pull/2004) and changing `case member` to `case member(conformances: [TypeSyntax])` would be API breaking since clients can no longer construct it as `["MyMemberMacro": .member]".

Reply:
In theory, we could manage that using static variables and default values for associated types, like this:

```swift
enum MacroRoleContext {
  case expression
  // ...
  case member(conformances: [String] = [])
  // ...
  case `extension`(conformances: [String])

  static var defaultMember: Self {
    return .member()
  }
}
```

By doing so, we could also gain a tool to inform users about API changes, by deprecation warnings on those static variables, that they can update their tests with the new functionalities.

Reply:
btw for the first time we been considering using such a context here https://github.com/apple/swift-syntax/pull/1859#discussion_r1245612254

Reply:
Unfortunately that’s not possible AFAICT

```swift
enum X {
  case foo(String)

  static var foo: X = .foo("a")
}

let a = X.foo // error: Ambiguous use of 'foo'
```

Reply:
lol. I'm confused right now. Why this: `let a: X = .foo` works and this: `let a = X.foo` doesn't? 🤯

![Uploading Screenshot 2023-11-09 at 8.19.41 PM.png…]()

Reply:
It doesn't look like a feature... it looks like a bug to me 😵‍💫

Reply:
Might be worth filing an issue in github.com/apple/swift about

Reply:
Yeah, I guess so. I've add it [here](https://github.com/apple/swift/issues/69772).

Reply:
> I’m not a fan of adding another parameter to `assertMacroExpansion`. What I would prefer is that the `macros` parameter becomes `[String: MacroSpec]` with
> 
> ```swift
> struct MacroSpec {
>   let type: MacroType
>   let conformances: [TypeSyntax]
> }
> ```
> 
> The benefit of that is that we could expand `MacroSpec` in the future without the need to have more parameters to `assertMacroExpansion`.
> 
> We still need to keep the current function that takes `macros: [String: Macro.Type]` around for API-compatibility.

@ahoppen made this change